src/core/queryObserver.ts

```diff
@@ -428,6 +428,9 @@ export class QueryObserver<
     const prevQueryResult = queryChange
       ? this.currentResult
       : this.previousQueryResult
+    const queryKey = prevQueryResult
```

Comment:
not sure I understand that. Why would we be interested in the key of the previous query?

Reply:
Without this we lost some memoization because the queryKey array was new each time the hook was called.

Said another way: prevQueryResult.queryKey !== query.queryKey (but they are shallow equal).

The only reason I caught that was because a test failed.

Reply:
Is it important to keep referential stability between queryKeys? If yes, we should likely triple equal compare the query hashes. 

What I don’t understand is how this can work with changing keys. For example: having `['todos', id]` as key and id transitions from 1 to 2. The previous key would be `['todos', 1]` and the new one is `['todos', 2]` - but with the logic above we would get the old, wrong key returned, wouldn’t we..?

Reply:
I need to read the code a little more closely, but it's my understanding that each time createResult on a QueryObserver is called, it's always the same "query" (that is, a unique queryKey [hash]) but the result is being updated with new information (retry, success, error, the hook being called again with no update, etc). Under this understanding, changed keys would result in two instances of a QueryObserver (one for id #1 and another for id #2 in your example above). Are QueryObservers pooled in the way you're suggesting, or is there one QueryObserver per unique queryHash?

I'm going to go read the code again and I'll check back.

Reply:
So I think this needs to change to:

`const queryKey = queryChange ? query.queryKey : prevQuery.queryKey`

but this is causing a test failure that I'm going to need to figure out. Help would be great. I think it has to do with the shallow result checking in updateResult ~L623 since the queryKey is an array not a primitive like everything else.